### PR TITLE
Java code generator uses the type annotations on the AST

### PR DESCRIPTION
**Pull request by jystic**
This is part of the refactoring to be done to make the C codegen easier to write.

### CODE REVIEW

File: test/cli/repl/run [hunk 1]
```diff
@@ -7,7 +7,7 @@ do
     echo " --- Test: $dir"
     if [ $update_expected -eq 1 ]
     then
-        dist/build/icicle-repl/icicle-repl ":import data/libs/prelude.icicle" ":load test/cli/repl/data.psv" < $dir/script | diff $dir/expected -
+        dist/build/icicle-repl/icicle-repl ":import data/libs/prelude.icicle" ":load test/cli/repl/data.psv" < $dir/script | diff -u $dir/expected -
```
Comment (amosr) [inline test/cli/repl/run]: does this give nicer output or something?

File: src/Icicle/Common/Type.hs [hunk 2]
```diff
@@ -272,6 +263,29 @@ instance Pretty FunType where
...
+    BoolT      -> text "Bool"
+    DateTimeT  -> text "DateTime"
+    StringT    -> text "String"
+    ArrayT t   -> parens' (text "Array " <>  ppSub t)
+    MapT k v   -> parens' (text "Map"    <+> ppSub k <+> ppSub v)
+    OptionT a  -> parens' (text "Option" <+> ppSub a)
+    PairT a b  -> parens  (ppTop a <> text ", " <> ppTop b)
+    StructT fs -> parens' (pretty fs)
+  where
+    parens' | needParens = parens
```
Comment (amosr) [inline src/Icicle/Common/Type.hs]: hooray!

**Comment by amosr** [comment]
yep, this is definitely worth it. thanks for convincing me to do it properly :seedling:

File: test/cli/repl/run (hunk 1 above)
Comment (jystic) [inline test/cli/repl/run]: Yeah it makes it the same as a `git diff` so you get a bit more context:

**file.a**
```
let x = read# [Int] ptr
let y = read# [Int] ptr

foreach (e : es) {
  printf# "f(x, y, e) = %d" (f x y e)
}

let z = read# [Int] ptr
```

**file.b**
```
let x = read# [Int] ptr
let y = read# [Int] ptr

foreach (e : es) {
  printf# "g(x, y, e) = %d" (g x y e)
}

let z = read# [Double] ptr
```

**diff file.a file.b**
```diff
5c5
<   printf# "f(x, y, e) = %d" (f x y e)
---
>   printf# "g(x, y, e) = %d" (g x y e)
8c8
< let z = read# [Int] ptr
---
> let z = read# [Double] ptr
```

**diff -u file.a file.b**
```diff
--- file.a	2015-08-28 05:11:08.000000000 +1000
+++ file.b	2015-08-28 05:11:12.000000000 +1000
@@ -2,7 +2,7 @@
 let y = read# [Int] ptr

 foreach (e : es) {
-  printf# "f(x, y, e) = %d" (f x y e)
+  printf# "g(x, y, e) = %d" (g x y e)
 }

-let z = read# [Int] ptr
+let z = read# [Double] ptr
```

**Pull request merged**